waddrmgr/scoped_manager.go

```diff
@@ -1636,6 +1639,117 @@ func (s *ScopedKeyManager) ImportScript(ns walletdb.ReadWriteBucket,
 	return scriptAddr, nil
 }
 
+// ImportWitnessScript imports a user-provided script into the address manager.  The
+// imported script will act as a pay-to-witness-script-hash address.
+//
+// All imported script addresses will be part of the account defined by the
+// ImportedAddrAccount constant.
+//
+// When the address manager is watching-only, the script itself will not be
+// stored or available since it is considered private data.
+//
+// This function will return an error if the address manager is locked and not
+// watching-only, or the address already exists.  Any other errors returned are
+// generally unexpected.
+func (s *ScopedKeyManager) ImportWitnessScript(ns walletdb.ReadWriteBucket,
+	script []byte, bs *BlockStamp) (ManagedScriptAddress, error) {
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	// The manager must be unlocked to encrypt the imported script.
+	if s.rootManager.IsLocked() && !s.rootManager.WatchOnly() {
+		return nil, managerError(ErrLocked, errLocked, nil)
+	}
+
+	// Prevent duplicates.
+	sh := sha256.Sum256(script)
```

Comment:
Combined with the comment above: we could extract this duplicated code into its own function that just takes a `baseScriptAddress` type (and the new cryptoKeyType parameter mentioned in the other comment).